### ISSUE TITLE
Is it possible to use COCO pretrained weights with my 3 class dataset?

### ISSUE THREAD

**Issue by dariogonle** [ultralytics/yolov5#5449]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I have a 3 classes datasets, being one of it the class Person. I would like to use the COCO pretrained weights to improve the results of my Person class.

The problem I see is that COCO has 80 classes and I only have 3. I'm having problems when I try to use the COCO pretrained weights with my dataset, and I was wondering if it is as simple as it cannot be done. If you have a 3 classes dataset you cannot use COCO pretrained weights. Is this correct?

Thank you in advance.

**Comment by Zengyf-CVer** [comment]
@dariogonle 
It has nothing to do with the number of categories in your custom data set. From your description, I guess that there is a problem with your custom data set. Carefully check the category assignment and instance assignment issues of the training set and validation set.

**Comment by Le0v1n** [comment]
Actually yes, you need modify some configurations and match the label requirements of YOLO v5.

**Comment by dariogonle** [comment]
@Zengyf-CVer, @Le0v1n or @glenn-jocher  Is it possible to train from COCO pretraiend weights but using only the class Person?

I mean, do a fine tuning of COCO weights with the class Person of my dataset.

**Comment by glenn-jocher** [comment]
@dariogonle YOLOv5 can train on any dataset using any YOLOv5 model as pretrained weights, class counts are irrelevant. Nothing special is needed, just run a training command:

i.e. train VOC.yaml (10 classes) starting from YOLOv5s.pt (80 classes):
```bash
python train.py --data VOC.yaml --weights yolov5s.pt
```

**Comment by dariogonle** [comment]
Thank you for your answer @glenn-jocher, but I am a bit confused.

If I use my own dataset.yaml
```
nc: 3
names: ['person', 'VE', 'HE']
```

I get the following results:
![image](https://user-images.githubusercontent.com/80393891/139852623-8699648e-43e2-44c0-9ab4-670eec96e070.png)

but if I change my dataset.yaml (the paths to the dataset are the same) to

```
nc: 80  # number of classes
names: ['person', 'bicycle', 'car', 'motorcycle', 'airplane', 'bus', 'train', 'truck', 'boat', 'traffic light',
        'fire hydrant', 'stop sign', 'parking meter', 'bench', 'bird', 'cat', 'dog', 'horse', 'sheep', 'cow',
        'elephant', 'bear', 'zebra', 'giraffe', 'backpack', 'umbrella', 'handbag', 'tie', 'suitcase', 'frisbee',
        'skis', 'snowboard', 'sports ball', 'kite', 'baseball bat', 'baseball glove', 'skateboard', 'surfboard',
        'tennis racket', 'bottle', 'wine glass', 'cup', 'fork', 'knife', 'spoon', 'bowl', 'banana', 'apple',
        'sandwich', 'orange', 'broccoli', 'carrot', 'hot dog', 'pizza', 'donut', 'cake', 'chair', 'couch',
        'potted plant', 'bed', 'dining table', 'toilet', 'tv', 'laptop', 'mouse', 'remote', 'keyboard', 'cell phone',
        'microwave', 'oven', 'toaster', 'sink', 'refrigerator', 'book', 'clock', 'vase', 'scissors', 'teddy bear',
        'hair drier', 'toothbrush']  # class names
```

I get the following:
![image](https://user-images.githubusercontent.com/80393891/139853084-64d525a2-6b73-44ca-8de8-12b5564cabe6.png)


In both cases I have used the following:
```
python train.py --epochs 1 --batch 8 --data dataset.yaml --hyp data/hyps/hyp.finetune.yaml --weights 'yolov5m.pt' --img-size 704 
```

So something weird must be happening, because with my dataset.yaml it is not working.

**Comment by glenn-jocher** [comment]
@dariogonle if your dataset is COCO then you should use the `nc` and `names` field you showed, if you have a custom dataset you supply your custom dataset `nc` and `names`.

Start from **Train Custom Dataset** tutorial:

## YOLOv5 Tutorials

* [Train Custom Data](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data)&nbsp; 🚀 RECOMMENDED
* [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/)&nbsp; ☘️ RECOMMENDED
* [Weights & Biases Logging](https://github.com/ultralytics/yolov5/issues/1289)&nbsp; 🌟 NEW
* [Supervisely Ecosystem](https://docs.ultralytics.com/yolov5/tutorials/model_export8)&nbsp; 🌟 NEW
* [Multi-GPU Training](https://docs.ultralytics.com/yolov5/tutorials/multi_gpu_training)
* [PyTorch Hub](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading)&nbsp; ⭐ NEW
* [TorchScript, ONNX, CoreML Export](https://docs.ultralytics.com/yolov5/tutorials/model_export) 🚀
* [Test-Time Augmentation (TTA)](https://docs.ultralytics.com/yolov5/tutorials/test_time_augmentation)
* [Model Ensembling](https://docs.ultralytics.com/yolov5/tutorials/model_ensembling)
* [Model Pruning/Sparsity](https://docs.ultralytics.com/yolov5/tutorials/model_pruning_and_sparsity)
* [Hyperparameter Evolution](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution)
* [Transfer Learning with Frozen Layers](https://docs.ultralytics.com/yolov5/tutorials/transfer_learning_with_frozen_layers)&nbsp; ⭐ NEW
* [TensorRT Deployment](https://github.com/wang-xinyu/tensorrtx)

**Comment by dariogonle** [comment]
> @dariogonle if your dataset is COCO then you should use the `nc` and `names` field you showed, if you have a custom dataset you supply your custom dataset `nc` and `names`.
> 
> Start from **Train Custom Dataset** tutorial:
> 
> ## YOLOv5 Tutorials
> * [Train Custom Data](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data)  🚀 RECOMMENDED
> * [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/)  ☘️ RECOMMENDED
> * [Weights & Biases Logging](https://github.com/ultralytics/yolov5/issues/1289)  🌟 NEW
> * [Supervisely Ecosystem](https://docs.ultralytics.com/yolov5/tutorials/model_export8)  🌟 NEW
> * [Multi-GPU Training](https://docs.ultralytics.com/yolov5/tutorials/multi_gpu_training)
> * [PyTorch Hub](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading)  ⭐ NEW
> * [TorchScript, ONNX, CoreML Export](https://docs.ultralytics.com/yolov5/tutorials/model_export) 🚀
> * [Test-Time Augmentation (TTA)](https://docs.ultralytics.com/yolov5/tutorials/test_time_augmentation)
> * [Model Ensembling](https://docs.ultralytics.com/yolov5/tutorials/model_ensembling)
> * [Model Pruning/Sparsity](https://docs.ultralytics.com/yolov5/tutorials/model_pruning_and_sparsity)
> * [Hyperparameter Evolution](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution)
> * [Transfer Learning with Frozen Layers](https://docs.ultralytics.com/yolov5/tutorials/transfer_learning_with_frozen_layers)  ⭐ NEW
> * [TensorRT Deployment](https://github.com/wang-xinyu/tensorrtx)

@glenn-jocher it is true that if I do it as you said, I can train a model with my dataset starting from COCO's pretrained weights, but I don't want to train a new model, what I want to do is a fine tunning, I mean get COCO's weights of the Person  class and fine tuning it during few epochs with my Person images.

My dataset is a custom one, but I want to use the pretrained weights of COCO, as they'd improve the Person class metrics . So my questions are the following: Shoud I use my own .yaml file?
```
nc: 3
names: ['person', 'VE', 'HE']
```

Should I use the following command to train?
```
python train.py --epochs 15 --batch 8 --data customDataset.yaml --hyp data/hyps/hyp.finetune.yaml --weights yolov5m.pt --img-size 704 
```

Because if I do it like this I get poor results, but in case I use coco.yaml file I get good ones.


Thank you in advance.

**Comment by dariogonle** [comment]
Ok, I think I got it.

Thank you all.

**Comment by Amulya2110** [comment]
Hi @dariogonle 
I am also working on similar problem, where-in i want to improve the performance of a "dog" class , by taking pre-trained coco weights as initial weights and train on top my custom dog data.

 By labelling my custom images with label 16 (dog class from pre-trained model) and by pointing to coco.yaml file and re-training with my custom images will help? Orelse  could you please let me know if there is any other way that you have figured out.

Thankyou

**Comment by dariogonle** [comment]
Hi @Amulya2110,

I can tell you what have I done.

My original customDataset.yaml was:
```
train: ../../Datasets/CHV/train/images
val: ../../Datasets/CHV/test/images

nc: 3
names: ['Person', 'VE', 'HE']
```

and I changed it to this:
```
train: ../../Datasets/CHV/train/images
val: ../../Datasets/CHV/test/images

nc: 80  # number of classes
names: ['Person', 'VE', 'HE', 'motorcycle', 'airplane', 'bus', 'train', 'truck', 'boat', 'traffic light',
        'fire hydrant', 'stop sign', 'parking meter', 'bench', 'bird', 'cat', 'dog', 'horse', 'sheep', 'cow',
        'elephant', 'bear', 'zebra', 'giraffe', 'backpack', 'umbrella', 'handbag', 'tie', 'suitcase', 'frisbee',
        'skis', 'snowboard', 'sports ball', 'kite', 'baseball bat', 'baseball glove', 'skateboard', 'surfboard',
        'tennis racket', 'bottle', 'wine glass', 'cup', 'fork', 'knife', 'spoon', 'bowl', 'banana', 'apple',
        'sandwich', 'orange', 'broccoli', 'carrot', 'hot dog', 'pizza', 'donut', 'cake', 'chair', 'couch',
        'potted plant', 'bed', 'dining table', 'toilet', 'tv', 'laptop', 'mouse', 'remote', 'keyboard', 'cell phone',
        'microwave', 'oven', 'toaster', 'sink', 'refrigerator', 'book', 'clock', 'vase', 'scissors', 'teddy bear',
        'hair drier', 'toothbrush']  # class names
```

And I just fine tunning it with the following command:
```
python train.py --epochs 15 --batch 8 --data customDataset.yaml --hyp data/hyps/hyp.finetune.yaml --weights yolov5m.pt --img-size 704
```

The result weights file will only have 3 classes as in my dataset there are only labels for classes 0, 1 and 2.

**Comment by Amulya2110** [comment]
@glenn-jocher @dariogonle 
Thankyou for the reply @dariogonle .
I understood what you have done , but i am unable to understand this point "The result weights file will only have 3 classes as in my dataset there are only labels for classes 0, 1 and 2.".

As the config file you pointed to while training says you have nc:80 , 80classes , so ideally shouldn't the weights file be of 80 class output even though in your custom dataset you have given labels for only 3 classes?

Does the weight file have no.of classes with respect to no.of classes labelled  in our custom data images or with respect to the nc defined in config file(data.yaml)?

**Comment by dariogonle** [comment]
At the beginning it is true that you're setting 80 class, but as in this case I only have 3 clases in my dataset (0, 1 and 2) the resulting weights will only have 3 classes.
